### PR TITLE
txnsync: avoid large allocation tie-ups

### PR DESCRIPTION
**Pull request by tsachiherman**
## Summary

The changes in this PR are focused around improving the memory utilization of the transaction sync, and handling excessive load on the transaction handler.

Improvements include:
1. The transaction sync allocate all the `transactions.SignedTxn` as a single allocated array. This works fine for the transaction sync, however, it create an inefficiently for the transaction pool : If only a single transaction group is going to be placed on the transaction pool, the entire backing array is still going to remain allocated. To resolve that, the transaction handler would re-allocate the transactions before placing them into the transaction pool.
2. The `solicitedAyncTxHandler` was refactored to contain a list of `*txGroups` instead of just `txGroups`. The rationale here is that the channel structure doesn't really clear out (internally) the `txGroups` object. In that case, a reference to that object remains in existence, and that prevent the GC from releasing the transaction groups array. Using a pointer, and clearing the content of the pointer allows us to ensure there is no additional reference to that piece of memory.
3. The `IncomingTransactionGroups` now have a way to report that it was not able to enqueue the transaction group array due to a full backlog channel. When that happens, the transaction sync is notified, allowing it to request no further transactions from its peer. Effectively, it allows a slow relay to recover from higher-then-supported message volume.
4. Reduce the frequency of the "transaction handler backlog is full" warning message. The message would now only be printed the first time it reaches that condition, and would not be printed again unless its state was reset by having the transaction handler backlog get mostly cleared.

## Test Plan

- [x] Use existing tests to ensure no regression.
- [x] Deploy an S1 network to verify functionality.

### CODE REVIEW

**Review by algonautshant — COMMENTED** [review]
Partial reviews.
I have once concern and one minor comment. 
Added a test (PR to this branch) for the new incoming addition.

File: data/txHandler.go [hunk 1]
```diff
@@ -381,6 +381,17 @@ func (handler *TxHandler) processDecodedArray(unverifiedTxGroups []transactions.
 	// so we can safely treat the transaction as a verified transaction.
 	verifiedTxGroup := unverifiedTxGroups
 
+	// before calling RememberArray we should reallocate the individual remaining
+	// signed transactions - these transactions were allocated in bulk by the
+	// transaction sync. By re-allocating the backing storage, we would allow the
+	// original backing storage ( which includes transactions that won't go intp the
```
Comment (algonautshant) [inline data/txHandler.go:387]: typo: intp -> into

**Review by algonautshant — COMMENTED** [review]
Looks great. 
One concern, and couple of minor issues.

File: txnsync/incoming.go [hunk 2]
```diff
@@ -265,12 +267,27 @@ MainForLoop:
 		// if we received at least a single transaction group, then forward it to the transaction handler.
 		if len(incomingMsg.transactionGroups) > 0 {
 			// send the incoming transaction group to the node last, so that the txhandler could modify the underlaying array if needed.
-			transacationPoolSize = s.node.IncomingTransactionGroups(peer, peer.nextReceivedMessageSeq-1, incomingMsg.transactionGroups)
+			currentTransacationPoolSize := s.node.IncomingTransactionGroups(peer, peer.nextReceivedMessageSeq-1, incomingMsg.transactionGroups)
+			// was the call reached the transaction handler queue ?
+			if currentTransacationPoolSize >= 0 {
+				// we want to store in transacationPoolSize only the first call to IncomingTransactionGroups
```
[the remark this answers is not in the document]
Comment (tsachiherman) [reply txnsync/incoming.go:273]: The call to IncomingTransactionGroups doesn't add the transactions to the transaction pool right away. It goes through a queue, and need to have its signature verified before it ends up on the transaction pool.
The returned size of the transaction pool here doesn't need to be accurate - we don't really care about the exact number of transactions; we just want to have a ballpark number to help us realize if we need to ramp up the messages rate.

[the remark this answers is not in the document]
Comment (algonautshant) [reply txnsync/incoming.go:273]: Even if the transactions are not added to the pool right away, they may be added during the iterations of the loop and counted  in currentTransacationPoolSize right? And later double counted in 
		`s.onTransactionPoolChangedEvent(MakeTranscationPoolChangeEvent(transacationPoolSize+totalAccumulatedTransactionsCount, transactionHandlerBacklogFull))
`

Anyway, this is fine if the accuracy is not important here.

**Review by algonautshant — APPROVED** [review]
Looks great. 
You can also merge my PR with the test addition before merging this.

**Pull request merged**